Integrations/src/main/java/io/deephaven/integrations/python/PythonDeephavenSession.java

```diff
@@ -308,8 +308,10 @@ public String scriptType() {
     @Override
     public Object unwrapObject(Object object) {
         if (object instanceof PyObject) {
-            final PyObject pyObject = (PyObject) object;
-            final Object unwrapped = module.unwrap_to_java_type(pyObject);
+            final Object unwrapped;
+            try (PyObject pyObject = (PyObject) object) {
```

Comment:
I'm not quite sure about this fix, as it risks the caller not wanting the unwrapped object to be closed. However, this is used in only a few places, and it seems safe to do this, and it prevents a leak. This is added in its own commit, and would be easy to remove, and doesn't affect the rest of the patch directly (except for its blatent wrongness).

There are many more remaining reference leaks in our code - I spent time auditing how often we create/close PyObjects, and the results are not great. Some of these leaks arent terrible, like references to boolean values, but we quickly leak lots of other values in terrible ways, and should find a better way to address this. 

We could consider adding a finalize() to jpy's PyObject, or drop Java8 support and look into Cnealer/etc instead, but probably only want to do that as a logging mechanism rather than actually managing resources this way.

Reply:
I don't think we should be closing PyObject here.

`finalize()` is a very critical time, and I don't think we want to risk blocking on the GIL (which is necessary when we decRef).

What is `Cnealer/etc`?

Reply:
Sorry, apparently typo'd [Cleaner](https://docs.oracle.com/javase/9/docs/api/java/lang/ref/Cleaner.html) in a weird way.

I'm not sure how finalize enters into this discussion - liveness should not depend on finalize at all, and LivenessArtifact and its hierarchy don't even have a finalize() override to ensure we unmanage when gc'd. Instead, these will almost all be called (with the exception of this one place I've noted might not belong in the patch) when the client calls SessionService.Release, when a session ends, or at the end of a submit() call's own liveness scope, etc.

Reply:
In this case, you mentioned

> We could consider adding a finalize() to jpy's PyObject

I'm not in favor of that. I think we do a lot of what Cleaner already does, see `org.jpy.PyObjectReferences#cleanupOnlyUseFromGIL(long[])`

Reply:
Agreed, I also don't want to add finalize to PyObject, but I do want to decref objects that are dropped by java. Example calling code that is fixed by this, in ScopeTicketResolver:
```
        T export = gss.getExecutionContext().getUpdateGraph().sharedLock().computeLocked(() -> {
            T scopeVar = null;
            try {
                // noinspection unchecked
                scopeVar = (T) gss.unwrapObject(gss.getVariable(scopeName));
            } catch (QueryScope.MissingVariableException ignored) {
            }
            return scopeVar;
        });
```
The result of getVariable immediately goes out of scope and isn't accessible again. We could store its result in a local, and test if it is Closable, and if so use try-with-resources? It seems a shame to need to do that.

Another option is to actually change the semantics of getVariable to return this new PyObjectRefCountedNode instead, and assume that a parent liveness scope exists. There is no explicit liveness scope specified here, but _most_ (didn't confirm all) calls that get here have an implicit/automatic liveness scope. However, changing getVariable seemed like too big of a step, though it would fix this specific case.

Other thoughts? Back this one piece out and plan on a bigger change later to update getVariable to always wrap the same way?

Reply:
Yes - I suggest backing this change out and adding a ticket. I think there are a few different ways we could tackle it in the future. Another option is new method, `getUnwrappedVariable`.